design-proposals/storage-migration/storage-class-migration.backup

```diff
@@ -0,0 +1,271 @@
+# Overview
+This design proposal extends KubeVirt APIs to enable the migration between storage classes.
+
+## Motivation
+In some circumstances, the guest owner may want to have the option of switching the storage provider or class for a group of volumes used by an active VM. The VM must currently be turned off for this operation due to service disruptions.
+
+
+## Goals
+Design and develop an API to handle storage migrations.
+
+ ## Non  Goals
+ This feature doesn't cover the VM Live migration with RWO volumes.
+ 
+ ## Definition of Users
+ This feature addresses for all KubeVirt users who want to move to another storage classes without workload interuptions.
+ 
+## User Stories
+  1. A storage class of volumes used by a running VM is deprecated
+  2. A more performant storage class is now available in the cluster and the    user wants start using it.
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD `StorageMigration`
+
+```golang
+type StorageMigration struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        Spec StorageMigrationSpec `json:"spec"`
+        Status *StorageMigrationStatus `json:"status,omitempty"`
+}
+
+// StorageMigrationList is a list of StorageMigration resources
+type StorageMigrationList struct {
+        metav1.TypeMeta `json:",inline"`
+        metav1.ListMeta `json:"metadata"`
+        Items []StorageMigration `json:"items"`
+}
+
+type MigratedVolume struct {
+        SourcePvc      string `json:"sourcePvc,omitempty" valid:"required"`
+        DestinationPvc string `json:"destinationPvc,omitempty" valid:"required"`
+}
+
+type MigrationStorageClass struct {
+        SourceStorageClass      string `json:"sourceStorageClass,omitempty" valid:"required"`
+        DestinationStorageClass string `json:"destinationStorageClass,omitempty" valid:"required"`
+}
+
+type ReclaimPolicySourcePvc string
+
+const (
+        DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+)
+
+// StorageMigrationSpec is the spec for a StorageMigration resource
+type StorageMigrationSpec struct {
+        VMIName string `json:"vmiName,omitempty" valid:"required"`
+        // MigratedVolumes is a list of volumes to be migrated
+        MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+        // MigrationStorageClass contains the information for migrationg the
+        // volumes of the source storage class to the destination storage class
+        MigrationStorageClass *MigrationStorageClass `json:"migrationStorageClass,omitempty"`
+        // ReclaimPolicySourcePvc describes how the source volumes will be
+        // treated after a successful migration
+        ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty"`
+}
+
+// StorageMigrationStatus is the status for a StorageMigration resource
+type StorageMigrationStatus struct {
```

Comment:
I think there should be conditions (and maybe a phase) to report that migration is running or not (stopped vm case)